File: src/0435-non-overlapping-intervals/solution.py
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        """
        Basically, we can either sort by start time or end time. Lets say
        we sort by start time.

        The interval to remove if overlapping is the interval that has the
        biggest/latest end time. This is because if we remove
        the earlier end time we will have to still remove another interval
        Like:
        [1, 3], [2, 5], [4, 6]
        if I remove (1, 3), I still have to remove (2, 5).
        So just remove (2, 5)

        Removal based on end time
        """
        intervals.sort(key=lambda x: x[0])
        curr_end = intervals[0][-1]
        removed_count = 0
        for i in range(1, len(intervals)):
            # Check if intervals do not merge
            start_time, end_time = intervals[i][0], intervals[i][-1]
            if start_time >= curr_end:
                # Update the most recent end time
                curr_end = end_time
            else:
                # Here we have to merge
                removed_count += 1
                # only select the smallest end time, removing the biggest end time
                curr_end = min(end_time, curr_end)
        return removed_count

File: src/0435-non-overlapping-intervals/solution.ts
function eraseOverlapIntervals(intervals: number[][]): number {
    /**
      * The interval to remove is the one with the smallest end timing
     */
     // sort ascending by start timings
     intervals.sort((a, b) => a[0] - b[0]);
     let currentEndTime = intervals[0].at(-1);
     let res = 0

     for (let i = 1; i < intervals.length; i++) {
        const [startTime, endTime] = intervals[i];

        // Check if can be placed infront
        if (currentEndTime <= startTime) {
            currentEndTime = endTime
        } else {
            res++;
            currentEndTime = Math.min(currentEndTime, endTime)
        }
     }
     return res;

};
